pkg/remediator/watch/filteredwatcher.go

```diff
@@ -504,14 +450,48 @@ func (w *filteredWatcher) shouldProcess(object client.Object) bool {
 	// its declaration. Otherwise we expect to get another event for the same
 	// object but with a matching GVK so we can actually compare it to its
 	// declaration.
-	if object.GetObjectKind().GroupVersionKind() != decl.GroupVersionKind() {
+	currentGVK := object.GetObjectKind().GroupVersionKind()
+	declaredGVK := decl.GroupVersionKind()
+	if currentGVK != declaredGVK {
+		klog.V(5).Infof("Received a watch event for object %q with kind %s, which does not match the declared kind %s. ",
+			id, currentGVK, declaredGVK)
 		return false
 	}
 
-	if !diff.CanManage(w.scope, w.syncName, object, diff.OperationManage) {
-		w.SetManagementConflict(object, commit)
+	if diff.CanManage(w.scope, w.syncName, object, diff.OperationManage) {
+		// TODO: Remove conflict error AFTER it has been resolved, not before.
+		w.conflictHandler.RemoveConflictError(id)
+		return true
+	}
+
+	desiredManager := declared.ResourceManager(w.scope, w.syncName)
+	currentManager, found := object.GetAnnotations()[metadata.ResourceManagerKey]
+	if !found || currentManager == "" {
+		// Should never happen, since CanManage returned false.
+		klog.Errorf("Reconciler %q received a watch event for object %q, which has no manager annotation, but CanManage returned false.",
+			desiredManager, id)
+		return false
+	}
+
+	// A root reconciler can override resource managed by namespace reconciler.
+	// It is not a conflict from the root reconciler's perspective.
+	// The namespace reconciler should detect and report the conflict.
+	if w.scope == declared.RootScope && !declared.IsRootManager(currentManager) {
+		klog.V(5).Infof("Reconciler %q received a watch event for object %q, which is managed by namespace reconciler %q. "+
+			"Root reconcilers are allowed to adopt objects managed by namespace reconcilers. "+
+			"The namespace reconciler should report the management conflict as an error.",
+			desiredManager, id, currentManager)
 		return false
```

Comment:
Shouldn't this return `true`, since the root reconciler is allowed to manage the object?

Reply:
If it gets here, CanManage is false, so no.

I'm not actually sure if this branch will ever execute, but if it does this prevents it from recording a conflict.

These branches were in SetManagementConflict, which didn't return anything. But if it got that far, shouldProcess was returning false.

Reply:
I think SetManagementConflict must have been called by some other code in the past, but it's not any more. Maybe it was from the applier code, which now adds conflicts to the conflict handler directly.

That said, the applier doesn't need either of these checks, and shouldn't actually use them. The object from the applier is the desired object, not the current object. So the currentManager here would be wrong. And the root applier should never report a management conflict event, because it always adopts.

Maybe we should just remove both checks as redundant.

Reply:
Removed